### PR TITLE
feat: conditional nav

### LINKED ISSUE #38808: Navigation Overhaul
In an effort to simplify and localize navigation we need to replace /news, /forum, and /learn nav items on the right side of the navigation with only one or two items conditionally. Here are the user stories:

- If users are signed in,  they should see two navigation items on the right side of the navigation:
Curriculum (goes to /learn)  and Portfolio( goes to the user's portfolio)

- if users are donating, they should see their picture in gold halo (refer to profile page for logic and styling) instead of the Portfolio item.

- If users are not signed in, they should see a call to action sign in button (refer to /news navigation donate button) instead of the Portfolio item.

Donation page specific:
- All users only see the centered logo with no nav items or search.

### PR DESCRIPTION
**Pull request by ahmaxed**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] All the files I changed are in the same world language, for example: only English changes, or only Chinese changes, etc.

Closes #38808

Here are the variations that could be considered:

### Navigation

Not logged in
Logged in no picture
Logged in no picture donating
Logged in no picture top contributor
Logged in no picture donating top contributor

Logged in with picture
Logged in with picture donating
Logged in with picture top contributor
Logged in with picture donating top contributor

Any of the above on the landing page

### Profile Page

 no picture
 no picture donating
 no picture top contributor
 no picture donating top contributor

with picture
with picture donating
with picture top contributor
with picture donating top contributor

### CODE REVIEW

**Review by raisedadead — CHANGES_REQUESTED** [review]
**Privacy Status:**

image

**User with default picture**

image

File: client/src/components/Header/components/NavLinks.js [hunk 1]
```diff
@@ -1,29 +1,39 @@
...
+
 const propTypes = {
-  displayMenu: PropTypes.bool
+  displayMenu: PropTypes.bool,
+  pathName: PropTypes.string.isRequired,
+  user: PropTypes.object
 };
 
-function NavLinks({ displayMenu }) {
+export function secondLink(user, pathName) {
```
Comment (raisedadead) [inline client/src/components/Header/components/NavLinks.js:14]: `secondLink` to me means there is a `firstLink` too? What does is aim to signify? Could it be something apt?

Comment (ahmadabdolsaheb) [reply client/src/components/Header/components/NavLinks.js:14]: the first link is the curriculum (/learn) that does not change. the second link is a conditional sign in/portfolio/avatar link.

I will make it more descriptive.

Comment (raisedadead) [reply client/src/components/Header/components/NavLinks.js:14]: I would just call it `authOrUserLink`, note we would have more links on the top nav with coderadio and others?

**Comment by ahmadabdolsaheb** [comment]
This should be ready to go.

File: client/src/components/Header/components/NavLinks.js [hunk 2]
```diff
@@ -1,29 +1,39 @@
...
+export function authOrUserLink(user, pathName) {
+  const isUserDonating = user && user.isDonating;
+  const isUserSignedIn = user && user.username;
+
+  if (pathName === '/' || !isUserSignedIn) {
+    return <Login data-test-label='landing-small-cta'> Sign In </Login>;
+  } else if (isUserDonating) {
+    return (
+      <Link className='supporter-nav-avatar' to={`/${user.username}`}>
+        {avatarRenderer(user.picture, user.username, 34, 1)}
```
Comment (raisedadead) [inline client/src/components/Header/components/NavLinks.js:23]:
```js
        {avatarRenderer(user.picture, user.username, 34, 1)}
```

Not a fan of the magic numbers here. We should be able to move this away into the component. This is clearly not dynamic because we are hard coding it. Does this need to be passed as a prop? Is there are better cleaner way.

**Review by raisedadead — CHANGES_REQUESTED** [review]

File: client/src/components/Header/components/NavLinks.js (hunk 2 above)
Comment (ahmadabdolsaheb) [reply client/src/components/Header/components/NavLinks.js:23]: I could move them to states, but it would not make any difference since the height of the nav is fixed and these numbers will not change.

This a similar convention to what we have here for rendering camper avatar.
 client/src/components/profile/components/Camper.js

Comment (raisedadead) [reply client/src/components/Header/components/NavLinks.js:23]: > these numbers will not change.

Exactly my point.

Since these values are fixed, it implies that they need to be set on a particular condition? This is an imperative approach and it could pretty straight to make it declarative. 

You already know what you want when the component needs to be placed in the nav, its needs to be 34px, with a padding of 1px it could be something else on a different component like on the profile page 256px with a padding of 5px.

Passing the css properties which are internal and native to the avatar component is imperative and breaks the notion of writing it as a component in the first place. It should be able to return the correct version of the component when asked by the nav or the profile page declaratively.

Comment (ahmadabdolsaheb) [reply client/src/components/Header/components/NavLinks.js:23]: Just to clarify, I would need to pass "profile" and "navigation" to avatar renderer and handle the sizes there?

Comment (raisedadead) [reply client/src/components/Header/components/NavLinks.js:23]: Is it okay if I push a change? I can demonstrate better with code.

Comment (ahmadabdolsaheb) [reply client/src/components/Header/components/NavLinks.js:23]: of course.

Comment (ahmadabdolsaheb) [reply client/src/components/Header/components/NavLinks.js:23]: @raisedadead, I am not sure I was able to understand your previous comment. If you tell me about the specific implementation you have in mind, I can make the changes accordingly. Thank you.

Comment (raisedadead) [reply client/src/components/Header/components/NavLinks.js:23]: Yikes, I forgot about this. 

We should do this declaratively, i.e something along the lines of

```js
{avatarRenderer(user.picture, user.username, 'sm')}
{avatarRenderer(user.picture, user.username, 'lg')}
```

the last param can be defined as `size` for example.

where the size='sm' can be equated to 34px width, 34px height, 1px padding
and the size='md' can be equated to 100px width, 100px height, 2px padding
and the size='lg' can be equated to 256px width, 256px height, 5px padding

This would make the component declative, with below uses. 

1. You can change the component later to whatever other internal styling without needing to change code all over the place.
2. You will not break the api of the `avatarRenderer`, should you forget to pass a param
3. You will be able to use the component elsewhere in pages if needed.

Basically you will be writing this as a 'component' vs 'implementation', the way it is right now makes it neither. There is a component but we are passing its implementation values.

Comment (ojeytonwilliams) [reply client/src/components/Header/components/NavLinks.js:23]: Could we make `avatarRenderer` a standard component?  Rather than passing in arguments, we just use props?  So

```
{avatarRenderer(user.picture, user.username, 'sm')}
```
becomes

```
<Avatar picture={user.picture} username={user.picture} size={'sm'}/>
```

or 

```
<Avatar user={user} size={'sm'}/>
```

and we `isRequired` the `picture` and `username` properties inside `Avatar`.  I think this is a more idiomatic way of going about it.

Comment (raisedadead) [reply client/src/components/Header/components/NavLinks.js:23]: Yup it should be a standard component.

**Comment by ahmadabdolsaheb** [comment]
There are a couple of considerations that I could use some feedback on.

1. the nav displays the sign-up by default and 'flashes' to portfolio when user data is available (hydration)
2. the are two 'show my profile' buttons. Should they be removed in this pr or removed later down the line?
3. should the donor images have a 'portfolio' tooltip on hover?

**Comment by ojeytonwilliams** [comment]
Sorry about that, Ahmad. Hopefully it should be fixed now.

**Comment by ahmadabdolsaheb** [comment]
@ojeytonwilliams, any thoughts on these?

> There are a couple of considerations that I could use some feedback on.
> 
> 1. the nav displays the sign-up by default and 'flashes' to portfolio when user data is available (hydration)
> 2. the are two 'show my profile' buttons. Should they be removed in this pr or removed later down the line?
> 3. should the donor images have a 'portfolio' tooltip on hover?
>

**Comment by raisedadead** [comment]
> the nav displays the sign-up by default and 'flashes' to portfolio when user data is available (hydration)

We should show a skeleton.

**Comment by ojeytonwilliams** [comment]
> 1. the nav displays the sign-up by default and 'flashes' to portfolio when user data is available (hydration)

We just need to track the `userFetchState`, I can do that if you like.

> 2. the are two 'show my profile' buttons. Should they be removed in this pr or removed later down the line?

on /learn?  There are two, but not for donating users, who would have their avatar in the top right.

> 3. should the donor images have a 'portfolio' tooltip on hover?

I don't quite follow.  What would this tooltip say?

**Comment by ahmadabdolsaheb** [comment]
> We just need to track the `userFetchState`, I can do that if you like.

Thanks, if you have something locally please push it. Otherwise, I can add that quickly.

> on /learn? There are two, but not for donating users, who would have their avatar in the top right.

There is one on map and one on settings. all users would have either "Profile" or their picture on the nav that takes them to their profile. Therefore I was wondering if all "go to my profile" buttons should be removed now.

> I don't quite follow. What would this tooltip say?
It might be obvious that the user's profile pic would take them to their profile. So I though on having a tooltip on hover with the word "Profile" for the image on navigation.

We have a similar functionality on news for the author image on the article cards.
image

**Comment by ojeytonwilliams** [comment]
Hi Ahmad, please ignore everything I said about profiles - I completely missed the fact the Avatar was inside a link.

I've pushed my code which should show the loader before switching to one of the options.  The page still re-flows, however.  I'm not sure how to avoid this, as we don't know in advance if we will show 'Sign In', 'Portfolio' or their avatar - all of which are different sizes.  Any ideas?

**Comment by ahmadabdolsaheb** [comment]
> Hi Ahmad, please ignore everything I said about profiles - I completely missed the fact the Avatar was inside a link.
> 
> I've pushed my code which should show the loader before switching to one of the options. The page still re-flows, however. I'm not sure how to avoid this, as we don't know in advance if we will show 'Sign In', 'Portfolio' or their avatar - all of which are different sizes. Any ideas?

I tried to make their sizes reasonably close. I will add extra space to make the amount of real state they take identical.

**Comment by raisedadead** [comment]
> I'm not sure how to avoid this, as we don't know in advance if we will show 'Sign In', 'Portfolio' or their avatar - all of which are different sizes.

Use skeletons.

**Comment by ojeytonwilliams** [comment]
Hey Ahmad, just one quick observation - my profile image is escaping its border:

image

**Comment by ojeytonwilliams** [comment]
Also they both, separately, link to my profile.  That is a little counter-intuitive - I would expect that each link should take me to different places.  Is it possible to either link to different things (profile and settings?) or combine them into a single link?

**Comment by ahmadabdolsaheb** [comment]
@o

> Also they both, separately, link to my profile. That is a little counter-intuitive - I would expect that each link should take me to different places. Is it possible to either link to different things (profile and settings?) or combine them into a single link?

@raisedadead would you like to link the profile image to settings?

**Comment by raisedadead** [comment]
Tested this. The two components should be wraped as one single element and take me to profile.

**Comment by ahmadabdolsaheb** [comment]
Thanks. Btw, I will go ahead and hide the whole navlinks behind a skeleton. That way I don't have to hardcode the margins and it would work well for internationalization.

**Comment by ahmadabdolsaheb** [comment]
Should be ready for review.
Test for donation state can be addressed after #39507 is resolved.

**Review by raisedadead — DISMISSED** [review]
Functionality LGTM. I will let Oliver take a look.
